build.rs

```diff
@@ -10,13 +10,17 @@ fn main() {
     if cfg!(feature = "development") {
         return;
     }
-    if let Ok(target) = env::var("TARGET") {
-        if !target.starts_with("arm-") {
```

Comment:
Some variant of this is still necessary to be able to automatically generate documentation without building the C library, but this check could be smarter.

Reply:
Yes, I agree, do you know of a way to get the compiler associated with a particular target in `build.rs`? Maybe the best would be requiring the user to specify it when run, e.g. `CC=arm-linux-gnueabihf-gcc cargo build --target=armv7-unknown-linux-gnueabihf`, since it can vary based on platform.

Reply:
Actually, it seems like the convention in rust is to use the [`gcc`](http://alexcrichton.com/gcc-rs/gcc/index.html) crate. I'll take a look a closer look at it when I have time.

Reply:
I would prefer it if we can still use the make files from wiringpi and wiringop. It would make upgrading them a bit easier if we don't have to care about any (smaller) changes in the build process... Actually these are the things I would want:

 * It should be easy to upgrade wiringpi and wiringop without having to know if their internal structure changes.
 * It should be possible to override the compiler by setting `CC` (see #16).
 * It should be possible to successfully run `cargo doc` on any platform.
 * It should be safe to accidentally run `cargo run` or even `sudo cargo run` on a development machine, without trashing the RAM or anything.

> Yes, I agree, do you know of a way to get the compiler associated with a particular target in `build.rs`?

Probably not, since there is not just one compiler per target. [These are the environment variables we get from Cargo](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts).

Reply:
> It should be easy to upgrade wiringpi and wiringop without having to know if their internal structure changes.

I'm not sure I agree, because there infinite many ways they could change their internal structure, and you can't cover all of the cases anyway, e.g. they could change to a completely different build system. We're pinning it to a particular version and we only use certain parts of the library, so I think this is fine.

> It should be possible to override the compiler by setting CC (see #16).

This is supported by `cc`. See the [docs](https://github.com/alexcrichton/cc-rs#external-configuration-via-environment-variables).

> It should be possible to successfully run cargo doc on any platform.
> It should be safe to accidentally run cargo run or even sudo cargo run on a development machine, without trashing the RAM or anything.

Not sure how these are relevant.

Reply:
> I'm not sure I agree, because there infinite many ways they could change their internal structure, and you can't cover all of the cases anyway, e.g. they could change to a completely different build system. We're pinning it to a particular version and we only use certain parts of the library, so I think this is fine.

What I mean is that running their make file will build the library the way they intended it to be built. I would otherwise have to inspect the make file to see if there is anything special going on.

> This is supported by cc. See the docs.

Good! :slightly_smiling_face: I just included it for completeness sake.

> Not sure how these are relevant.

Being able to run `cargo doc` is absolutely necessary for generating [the documentation at docs.rs](https://docs.rs/wiringpi/0.2.2/wiringpi/). They are running a script that just generates it for a handful of platforms and none of them are arm platforms. That's why this check we are commenting on existed in the first place.

As for the rest, wiringpi does a bunch of memory mapping and volatile writes to control the pins and I have no idea what happens if it would run on something that's not a Raspberry Pi. That's why I haven't even bothered making it compile properly for other platforms. Things throwing loud errors are pretty good indicators of not compiling for the correct architecture.

Reply:
> What I mean is that running their make file will build the library the way they intended it to be built. I would otherwise have to inspect the make file to see if there is anything special going on.

Yes, I got that. I just think if something special is indeed going on, we need take care of it explicitly anyway, on a case-by-case basis. Eg. they might split the library into multiple artifacts, how do you handle that without making changes to our build process?

> Being able to run cargo doc is absolutely necessary for generating the documentation at docs.rs. They are running a script that just generates it for a handful of platforms and none of them are arm platforms. That's why this check we are commenting on existed in the first place.

> As for the rest, wiringpi does a bunch of memory mapping and volatile writes to control the pins and I have no idea what happens if it would run on something that's not a Raspberry Pi. That's why I haven't even bothered making it compile properly for other platforms. Things throwing loud errors are pretty good indicators of not compiling for the correct architecture.

Oh, I see now :smile: I added the check back in.

Reply:
> Yes, I got that. I just think if something special is indeed going on, we need take care of it explicitly anyway, on a case-by-case basis. Eg. they might split the library into multiple artifacts, how do you handle that without making changes to our build process?

There's always a risk of changes being made, but it's more probable that the way the make file is used will be roughly the same, but what do I know? I see that there is an installation script there now, so it's already non-standard. Let's try it your way and see how it works.